File: apps/website/internal/services/httpretry.go
package services

import (
	"net/http"
	"strconv"
	"time"
)

// Discord 429 retry policy (T-130.2 F3-01): bounded attempts, honoring the
// Retry-After header (Discord sends fractional seconds). The wait is clamped so a
// hostile/global rate-limit response cannot park a request past the HTTP client's
// own 10s timeout.
const (
	max429Attempts    = 3
	default429Backoff = time.Second
	max429Backoff     = 5 * time.Second
)

// parseRetryAfter converts a Retry-After header value (seconds, possibly
// fractional) into a bounded wait, falling back to default429Backoff when the
// header is absent or malformed.
func parseRetryAfter(v string) time.Duration {
	secs, err := strconv.ParseFloat(v, 64)
	if err != nil || secs < 0 {
		return default429Backoff
	}
	d := time.Duration(secs * float64(time.Second))
	if d > max429Backoff {
		return max429Backoff
	}
	return d
}

// doWithRetryOn429 executes req via client, retrying up to max429Attempts times
// while the response is 429 Too Many Requests. Replayable bodies (req.GetBody —
// set by net/http for strings/bytes readers) are rewound between attempts. The
// final 429 response is returned to the caller, whose normal non-2xx handling
// surfaces it; waits abort early if the request context ends.
func doWithRetryOn429(client *http.Client, req *http.Request) (*http.Response, error) {
	for attempt := 1; ; attempt++ {
		if attempt > 1 && req.GetBody != nil {
			body, err := req.GetBody()
			if err != nil {
				return nil, err
			}
			req.Body = body
		}
		resp, err := client.Do(req)
		if err != nil {
			return nil, err
		}
		if resp.StatusCode != http.StatusTooManyRequests || attempt == max429Attempts {
			return resp, nil
		}
		wait := parseRetryAfter(resp.Header.Get("Retry-After"))
		//nolint:errcheck // best-effort: the 429 body is discarded before the retry.
		resp.Body.Close()
		select {
		case <-time.After(wait):
		case <-req.Context().Done():
			return nil, req.Context().Err()
		}
	}
}

File: apps/website/src/services/http_retry.rs
//! Bounded 429 retry — Rust port of `services/httpretry.go`. Shared by the Discord
//! client + the announcement webhook. Retries honor `Retry-After` (fractional
//! seconds), clamped so a hostile rate-limit can't park a request indefinitely.

use std::time::Duration;

use reqwest::{RequestBuilder, Response, StatusCode};

const MAX_429_ATTEMPTS: u32 = 3;
const DEFAULT_429_BACKOFF: Duration = Duration::from_secs(1);
const MAX_429_BACKOFF: Duration = Duration::from_secs(5);

/// Send `build()`'s request, retrying up to 3 times while the response is 429.
/// The request is rebuilt each attempt (fresh body). The final 429 is returned to
/// the caller, whose normal non-2xx handling surfaces it.
pub async fn send_with_retry_on_429<F>(build: F) -> reqwest::Result<Response>
where
    F: Fn() -> RequestBuilder,
{
    let mut attempt = 1;
    loop {
        let resp = build().send().await?;
        if resp.status() != StatusCode::TOO_MANY_REQUESTS || attempt == MAX_429_ATTEMPTS {
            return Ok(resp);
        }
        let wait = parse_retry_after(
            resp.headers()
                .get("retry-after")
                .and_then(|v| v.to_str().ok()),
        );
        attempt += 1;
        tokio::time::sleep(wait).await;
    }
}

/// Convert a `Retry-After` value (seconds, possibly fractional) into a bounded wait.
pub(crate) fn parse_retry_after(v: Option<&str>) -> Duration {
    match v.and_then(|s| s.parse::<f64>().ok()) {
        Some(secs) if secs >= 0.0 => {
            let d = Duration::from_secs_f64(secs);
            if d > MAX_429_BACKOFF {
                MAX_429_BACKOFF
            } else {
                d
            }
        }
        _ => DEFAULT_429_BACKOFF,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn retry_after_parsing_and_clamp() {
        assert_eq!(parse_retry_after(Some("2")), Duration::from_secs(2));
        assert_eq!(parse_retry_after(Some("0.5")), Duration::from_millis(500));
        assert_eq!(parse_retry_after(Some("100")), MAX_429_BACKOFF); // clamped
        assert_eq!(parse_retry_after(None), DEFAULT_429_BACKOFF);
        assert_eq!(parse_retry_after(Some("garbage")), DEFAULT_429_BACKOFF);
    }
}
